src/oneD/refine.cpp

```diff
@@ -77,153 +80,154 @@ int Refiner::analyze(size_t n, const double* z, const double* x)
             for (size_t j = 0; j < n; j++) {
-                v[j] = value(x, i, j);
+                val[j] = value(x, i, j);
             }
 
-            // slope of component i
+            // slope of component i (using forward difference)
             for (size_t j = 0; j < n-1; j++) {
-                s[j] = (value(x, i, j+1) - value(x, i, j))/(z[j+1] - z[j]);
+                slope[j] = (val[j+1] - val[j]) / dz[j];
             }
 
-            // find the range of values and slopes
-            double vmin = *min_element(v.begin(), v.end());
-            double vmax = *max_element(v.begin(), v.end());
-            double smin = *min_element(s.begin(), s.end());
-            double smax = *max_element(s.begin(), s.end());
-
-            // max absolute values of v and s
-            double aa = std::max(fabs(vmax), fabs(vmin));
-            double ss = std::max(fabs(smax), fabs(smin));
-
-            // refine based on component i only if the range of v is
-            // greater than a fraction 'min_range' of max |v|. This
-            // eliminates components that consist of small fluctuations
-            // on a constant background.
-            if ((vmax - vmin) > m_min_range*aa) {
-                // maximum allowable difference in value between adjacent
-                // points.
-                double dmax = m_slope*(vmax - vmin) + m_thresh;
+            // find the range of values and slopes of component i over the domain
+            double valMin = *min_element(val.begin(), val.end());
+            double valMax = *max_element(val.begin(), val.end());
+            double slopeMin = *min_element(slope.begin(), slope.end());
+            double slopeMax = *max_element(slope.begin(), slope.end());
+
+            // max absolute values of val and slope
+            double valMagnitude = std::max(fabs(valMax), fabs(valMin));
+            double slopeMagnitude = std::max(fabs(slopeMax), fabs(slopeMin));
+
+            // refine based on component i only if the range of val is greater than a
+            // fraction 'min_range' of max |val|. This eliminates components that
+            // consist of small fluctuations around a constant value.
+            if (valMax - valMin > m_min_range*valMagnitude) {
+                // maximum allowable difference in value between adjacent points. Based
+                // on the global min and max values of the component over the domain.
+                double max_change = m_slope*(valMax - valMin);
                 for (size_t j = 0; j < n-1; j++) {
-                    double r = fabs(v[j+1] - v[j])/dmax;
-                    if (r > 1.0 && dz[j] >= 2 * m_gridmin) {
-                        m_loc.insert(j);
-                        m_c.insert(name);
+                    // m_thresh prevents two adjacent points from being considered
+                    // if they are too close together.
+                    double ratio = fabs(val[j+1] - val[j]) / (max_change + m_thresh);
```

Comment:
See previous comment about `m_thresh`.

Reply:
Am I correct in my interpretation that this quantity is not a generic absolute tolerance. It's specifically the absolute tolerance for the solution component "closeness". In the m_curve section in refine.cpp we see that this tolerance is not used to prevent slope values below the tolerance, but rather an effective tolerance, relative to the solution components is computed.